### ISSUE TITLE
Clarify what Record.Value contains for TXT records

### ISSUE THREAD

**Issue by emersion** [libdns/libdns#107]
TXT records can contain multiple strings per record (see e.g. https://pkg.go.dev/github.com/miekg/dns#TXT). How should this be reflected in `Record`?

Some providers seem to ignore this fact and dump the first TXT string in `Value`. Some other providers seem to quote TXT values before populating `Value`.

This causes some interoperability issues between providers. For some providers, `a b c` will be interpreted as 3 TXT values, for some others it will be one TXT value. Spaces can be common for TXT, e.g. `v=DMARC1; p=none` is a typical value. TXT values can contain any character in theory, including `"`.

(There is a more general problem of what `Value` means for more complicated records like SOA, but let's keep this for later.)

**Comment by mholt** [comment]
Is there a semantic difference between two TXT records with values `a` and `b`, and one TXT record with two values, `a` and `b`?

**Comment by emersion** [comment]
Yes, there is a difference. See e.g. https://datatracker.ietf.org/doc/html/rfc6763#section-6.6 where the fields must be specified as separate strings, and each field may contain spaces in it.

Another example is https://datatracker.ietf.org/doc/html/rfc7489#section-6.1, where the strings are joined together without any kind of separator.

I don't think splitting into multiple records makes sense, because there are also semantic differences between one TXT record with 3 strings and 3 TXT records with 1 string each. Moreover, it's not clear what should happen with `Record.ID`.

**Comment by emersion** [comment]
Here is an example of a spec allowing multiple TXT strings in a single record, but not multiple TXT records: https://datatracker.ietf.org/doc/html/rfc7208#section-3.2.

And here is an example of a spec with multiple TXT strings per records, and multiple TXT records per name: https://datatracker.ietf.org/doc/html/rfc6763#section-6.8. Each TXT record describes a different variant of the same service (e.g. with a different version), and each TXT record contains one TXT string per key/value pair.

Also, note that a single TXT string is limited to 255 characters (which is why some specs like DKIM require users to concatenate the strings together without a separator: 255 characters are not enough to store cryptographic keys).

**Comment by mholt** [comment]
(These are good questions. Sorry for my delayed response, the Caddy 2.7 release has been keeping me busy. This is still on my list though and I'll get back soon!)

**Comment by emersion** [comment]
Sure, no worries at all!

If I had to think of a solution to this, it'd be one of:

1. Decide that we only support a single TXT value. That means that we shut the door at some use-cases (arguably niche).
2. Introduce a new struct field which can hold a slice of strings. This would lean in an API design where we add more fields specific to each record type (we already have `Priority`, we'd add more). Taken to the extreme, this results in one separate Go type per record type as the miekg/dns package does (and e.g. A/AAAA records hold a `net.IP` value instead of a string). The API type safety is increased but our API surface grows quite a bit. We don't necessarily need to go for the extreme, we could go for a middle-ground where `Value` stays a string and we add various additional record type specific fields, although it could become a bit complicated to tell which field can be used for which record type (e.g. `TXTValues` must be used instead of `Value` for TXT records).
3. Decide that the `Value` holds the raw, zonefile-encoded payload of the record. This means less type safety and more work for both providers and consumers if they want/need to decode the value. For instance marshalling and unmarshalling quoted TXT strings isn't super trivial, and each record has its own encoding.
4. Auto-magically detect whether the format of `Value` is quoted or not. If not quoted, treat as a single TXT string. If quoted, treat as potentially multiple TXT strings. This is pretty fragile and error-prone, especially if each and every provider needs to properly implement this.

**Comment by emersion** [comment]
Hi! Did you have the chance to look at the options listed above @mholt?

**Comment by mholt** [comment]
So sorry. Shortly thereafter we had a hospitalization in the family and then a newborn 😅  I haven't forgotten about this but have about 75-80 issues/PRs in my backlog still...

DNS is so messy... sounds like something I'd have designed. 😇 Anyway, I _kind of like_ option 2.

Is there an option 5? What if each libdns.Record contains exactly 1 TXT record and if there are multiple, we leave it to the provider implementations to do the concatenation if required by that provider. That way each TXT record can still have its own ID but can also be concatenated... I think?

**Comment by mholt** [comment]
@John-K Some providers attach IDs to individual records. That can be useful if the provider supports multiple TXT.

I suppose if the provider requires concatenation, then inputting separate records is lossy: multiple TXT go in, one TXT comes out. But that's still fine, since apparently concatenated TXT is semantically effectively the same as separate TXT?

If you still need metadata, what if we add a Metadata or Meta (or something) field to the `Record` struct? That way it could stay with the Record. Unless, of course, you only need the data for a certain transaction/lifecycle, then maybe a context value makes more sense. What do you think?

**Comment by emersion** [comment]
When I said "split"/"concatenate" above, I was only talking about doing that with a single record only, never to split into multiple records or concatenate multiple records into one.

An example of what I mean: let's say the user passes in a single TXT record with a value of 300 characters. A provider might decide to "split" that value into two strings (first with 255 characters, second with 45 characters) and store these two strings into a single TXT record. When fetching records from the provider, the provider might decide to "concatenate" these two strings contained in the single TXT record back into the 300 characters value.

At no time the TXT record gets split into 2 records, or 2 records get concatenated into a single one. This is only about the fact that TXT records hold a _list_ of strings of 255 characters max in the DNS RFC (not a single string).

**Comment by gucci-on-fleek** [comment]
Oh wow, I didn't notice this until just now. We discussed this in https://github.com/libdns/libdns/pull/153#discussion_r2018060325 and we ultimately decided on your option 1:

> 1. Decide that we only support a single TXT value. That means that we shut the door at some use-cases (arguably niche). When building a `Record` in a provider, if there are multiple TXT records, they get concatenated together in a single string without any separator. When reading a `Record` in a provider, if the string is more than 255 bytes long, it gets split into multiple TXT records.

**Comment by mholt** [comment]
PS. The [new godoc](https://pkg.go.dev/github.com/libdns/libdns@v1.0.0-beta.1) looks really good! Nice work!!
